File: typescript/src/test/typescript/movierental/customer-builder.ts
import {Customer} from "../../../main/typescript/movierental/customer";
import {Rental} from "../../../main/typescript/movierental/rental";

const NAME = "Roberts";

export class CustomerBuilder {
    private name: string = NAME;
    private rentals: Rental[] = [];

    public build(): Customer {
        const result = new Customer(this.name);
        for (const rental of this.rentals) {
            result.addRental(rental);
        }
        return result;
    }

    public withName(name: string): CustomerBuilder {
        this.name = name;
        return this;
    }

    public withRentals(...rentals: Rental[]): CustomerBuilder {
        this.rentals.push(...rentals);
        return this;
    }
}

File: scala/src/test/scala/movierental/CustomerBuilder.scala
package movierental

import scala.collection.mutable

class CustomerBuilder {
  var name: String = CustomerBuilder.NAME
  val rentals: mutable.MutableList[Rental] = mutable.MutableList[Rental]()

  def build(): Customer = {
    val result = new Customer(name)
    for (each <- rentals) {
      result.addRental(each)
    }
    return result
  }

  def withName(name: String): CustomerBuilder = {
    this.name = name
    return this
  }

  def withRentals(rentals: Rental*): CustomerBuilder = {
    this.rentals ++= rentals
    return this
  }
}

object CustomerBuilder {
  val NAME = "Roberts"
}